iohub/ngff_utils.py

```diff
@@ -0,0 +1,341 @@
+    kwargs : dict
+        Additional arguments to pass to the CZYX function
+    """
+
+    # TODO: temporary fix to slumkit issue
+    if _is_nested(input_channel_indices):
+        input_channel_indices = [
+            int(x) for x in input_channel_indices if x.isdigit()
+        ]
+    if _is_nested(output_channel_indices):
+        output_channel_indices = [
+            int(x) for x in output_channel_indices if x.isdigit()
+        ]
+
+    # Check if t_idx_in should be added to the func kwargs
+    # This is needed when a different processing is needed for each time point, for example during stabilization
+    all_func_params = inspect.signature(func).parameters.keys()
+    if "t_idx_in" in all_func_params:
+        kwargs["t_idx_in"] = t_idx_in
+
+    # Process CZYX given with the given indeces
+    # if input_channel_indices is not None and len(input_channel_indices) > 0:
+    click.echo(f"Processing t={t_idx_in} and channels {input_channel_indices}")
+    czyx_data = position.data.oindex[t_idx_in, input_channel_indices]
+    if not _check_nan_n_zeros(czyx_data):
+        transformed_czyx = func(czyx_data, **kwargs)
+        # Write to file
+        with open_ome_zarr(output_path, mode="r+") as output_dataset:
+            output_dataset[0].oindex[
+                t_idx_out, output_channel_indices
+            ] = transformed_czyx
+        click.echo(
+            f"Finished Writing.. t={t_idx_in} and channel output={output_channel_indices}"
+        )
+    else:
+        click.echo(f"Skipping t={t_idx_in} due to all zeros or nans")
+
+
+# TODO: modify how we get the time and channesl like recOrder (isinstance(input, list) or instance(input,int) or all)
+def process_single_position(
+    func,
+    input_data_path: Path,
+    output_path: Path,
+    time_indices_in: list = "all",
+    time_indices_out: list = [],
+    input_channel_idx: list = [],
+    output_channel_idx: list = [],
+    num_processes: int = mp.cpu_count(),
+    **kwargs,
+) -> None:
+    """
+    Register a single position with multiprocessing parallelization over T and C
+
+    Parameters
+    ----------
+    func : CZYX -> CZYX function
+        The function to be applied to the data
+    input_data_path : Path
+        The path to input position
+    output_path : Path
+        The path to output OME-Zarr Store
+    time_indices_in : list
+        The time indices to process.
+        Must match time_indices_out if not "all"
+    time_indices_out : list
+        The time indices to write to.
+        Must match time_indices_in if not empty.
+        Typically used for stabilization, which needs a per timepoint processing
+    input_channel_idx : list
+        The channel indices to process.
+        If empty list, process all channels.
+        Must match output_channel_idx if not empty
+    output_channel_idx : list
+        The channel indices to write to.
+        If empty list, write to all channels.
+        Must match input_channel_idx if not empty
```

Comment:
What are acceptable values here? We should provide examples. See https://numpy.org/doc/stable/user/basics.indexing.html. I think acceptable values should be a list of slices, e.g. [slice(0, 2), slice(2, 4), ...], or list of lists, e.g. [[0], [1], ...]. In both cases data[input_channel_idx[0]] would return a CZYX array, i.e. single dimensions are not dropped. We should not allow for list of integers, e.g. [0, 1, ...] as data[input_channel_idx[0]] would return a ZYX array which is not compatible with downstream processing and you need to do data[c_idx:c_idx+1]. Also, we should provide examples on how 5 channels can be turned into 3 channels during biref recon, for example. In that case, the input and output channel indices would look like [[0, 1, 2, 3, 4]] and [[0, 1, 2]], correct? This syntax then allow for easy multiprocessing along members of zip(input_channel_idx, output_channel_idx).

Reply:
Still need to test this but I added the docstring and var types.

Reply:
I think we should revisit your point here, @ieivanov.

I think @ieivanov has a case in mind where we need to apply the same transform to multiple groups of channels within the same CZYX array. For example, applying the same reconstruction to several groups of channels (i.e. a birefringence recon for the first five channels and the last five channels). Here @ieivanov is suggesting changes that would allow us to multiprocess across these reconstructions. 

For reconstructions I think this case happens quite rarely; each channel needs its own reconstruction. @ieivanov do you have different processing steps in mind?

We do apply multiple reconstructions to the same CZYX array (fluorescence to the first channel, birefringence to the last 5), but I currently handle this on the recOrder side separate reconstruction calls. Admittedly, these reconstructions could be faster through multiprocessing, but so far this is very rarely the bottleneck. 

Unless we have a good grouped multi-channel problem, I suggest we change back to `input/output_channel_indices : list | slice`. 

P.S. I do want `list`, not `int`, so that we can keep the shapes consistent as @ieivanov suggests. So for a single-channel process we should pass `input/output_channel_indices = [0]`.

Reply:
All addressed in person.